src/query/api/v1/handler/prometheus/remote/read.go

```diff
@@ -259,103 +260,119 @@ type ReadResult struct {
+			fmt.Errorf("cannot parse params: no expr"),
+			http.StatusBadRequest)
+	}
 
-		queryEnd, err := util.ParseTimeString(r.FormValue("end"))
-		if err != nil {
-			return nil, nil, xhttp.NewParseError(err, http.StatusBadRequest)
-		}
+	queryStart, err := util.ParseTimeString(r.FormValue("start"))
+	if err != nil {
+		return nil, xhttp.NewParseError(err, http.StatusBadRequest)
+	}
+
+	queryEnd, err := util.ParseTimeString(r.FormValue("end"))
+	if err != nil {
+		return nil, xhttp.NewParseError(err, http.StatusBadRequest)
+	}
 
-		req = &prompb.ReadRequest{}
+	req = &prompb.ReadRequest{}
+	expr, err := promql.ParseExpr(exprParam)
+	if err != nil {
+		return nil, xhttp.NewParseError(err, http.StatusBadRequest)
+	}
 
-		expr, err := promql.ParseExpr(exprParam)
-		if err != nil {
-			return nil, nil, xhttp.NewParseError(err, http.StatusBadRequest)
-		}
+	var (
+		evalRange time.Duration
+	)
 
+	promql.Inspect(expr, func(node promql.Node, path []promql.Node) error {
 		var (
-			evalRange time.Duration
+			start = queryStart
+			end   = queryEnd
 		)
-		promql.Inspect(expr, func(node promql.Node, path []promql.Node) error {
-			var (
-				start = queryStart
-				end   = queryEnd
-			)
-
-			switch n := node.(type) {
-			case *promql.VectorSelector:
-				if evalRange > 0 {
-					start = start.Add(-1 * evalRange)
-					evalRange = 0
-				}
-
-				if n.Offset > 0 {
-					offsetMilliseconds := time.Duration(n.Offset) * time.Millisecond
-					start = start.Add(-1 * offsetMilliseconds)
-					end = end.Add(-1 * offsetMilliseconds)
-				}
 
-				matchers, err := toLabelMatchers(n.LabelMatchers)
-				if err != nil {
-					return err
-				}
+		switch n := node.(type) {
+		case *promql.VectorSelector:
+			if evalRange > 0 {
+				start = start.Add(-1 * evalRange)
+				evalRange = 0
+			}
 
-				query := &prompb.Query{
-					StartTimestampMs: storage.TimeToPromTimestamp(start),
-					EndTimestampMs:   storage.TimeToPromTimestamp(end),
-					Matchers:         matchers,
-				}
+			if n.Offset > 0 {
+				offset := time.Duration(n.Offset)
+				start = start.Add(-1 * offset)
+				end = end.Add(-1 * offset)
+			}
 
-				req.Queries = append(req.Queries, query)
+			matchers, err := toLabelMatchers(n.LabelMatchers)
+			if err != nil {
+				return err
+			}
 
-			case *promql.MatrixSelector:
-				evalRange = n.Range
-				if evalRange > 0 {
-					start = start.Add(-1 * evalRange)
-					evalRange = 0
-				}
+			query := &prompb.Query{
+				StartTimestampMs: storage.TimeToPromTimestamp(start),
+				EndTimestampMs:   storage.TimeToPromTimestamp(end),
+				Matchers:         matchers,
+			}
 
-				if n.Offset > 0 {
-					offsetMilliseconds := time.Duration(n.Offset) * time.Millisecond
-					start = start.Add(-1 * offsetMilliseconds)
-					end = end.Add(-1 * offsetMilliseconds)
-				}
+			req.Queries = append(req.Queries, query)
 
-				matchers, err := toLabelMatchers(n.LabelMatchers)
-				if err != nil {
-					return err
-				}
+		case *promql.MatrixSelector:
```

Comment:
Can we reuse code between the matrix selector and vector selector somehow?  They're very similar.